ops/relation.py

```diff
@@ -434,6 +434,12 @@ def _validate_provider(self, event):
         logger.debug("Validating provider(s) : %s", event)
         consumed = self.consumes
 
+        if self.name not in self.framework.model.relations:
```

Comment:
Perhaps I have misunderstood the issue here. But this feels like fixing a possible bug in the Juju event model by a patch in the Operator Framework. Here is why I think so. A charm's upgrade hook handler should only be invoked when it has exactly the same environment it was running in, except for the fact it's own code is now updated. This must be so because the express intent of the upgrade event handler is to deal with only this situation and nothing else. If the environment (including relations) prior to upgrade is not created before invoking the charm's upgrade handler then Juju has violated its contract with the charm code. Perhaps this calls for a fix in the Operator Framework but is this the right thing to do ? Maybe as a stop gap arrangement. But perhaps we should raise this with Juju developers to ensure we do have the right event model works as expected and Juju honours it's contracts.  

Note that in creating the environment that existed prior to upgrade Juju is not obliged to (and perhaps should not) fire and relation events, unless of course there have actually been some changes to relations.

Reply:
I don't think it's a juju thing.
In my charm's constructor I have the following:
```python
        # charm code
        self.karma_lib = KarmaConsumer(
            self,
            "karma_dashboard",
            consumes={"karma": ">=0.86"},
        )
```
This code is run every time, even if there isn't a relation to karma in place (currently or ever).
See my suggestion [here](https://github.com/canonical/operator/pull/577#discussion_r678703381).

Reply:
This is exactly a stopgap so @sed-i can move forward with testing upgrades. I'm looking for a root cause. If `CharmMeta` is not appropriately populated on upgrades, it may explain the other "missing" lifecycle events also. I'll update here.

Reply:
I had read your comment when I wrote the above.